### ISSUE TITLE
Why fail fast on dial when connection limit reached but wait for a spot for incoming connections?

### ISSUE THREAD

**Issue by kdeme** [status-im/nim-libp2p#548]
I would like to understand the rationale why libp2p connection manager makes a dial fail fast when the maximum allowed peers is reached, while for incoming connections it can await a free slot. 
https://github.com/status-im/nim-libp2p/blob/master/libp2p/connmanager.nim#L423 versus https://github.com/status-im/nim-libp2p/blob/master/libp2p/connmanager.nim#L446

Would it not be better to allow for both to wait? And that way allow the "user" to cancel the dial when it waited long enough.

The current situation makes it favour incoming connections basically.

I realize this can be resolved by splitting limits over the `inSema` and `outSema`, but still, I think it would be nice to have the waiting behaviour also for outgoing connections.

Additionally incoming connections allow to be keep added up waiting (not added to the peer count though luckily) because they wait on yet another semaphore here: https://github.com/status-im/nim-libp2p/blob/master/libp2p/switch.nim#L324
Would it not be better to let these fail fast?

[That](https://github.com/status-im/nim-libp2p/blob/master/libp2p/switch.nim#L316) logic should probably also go in the connection management as now it is separated and not so visible, yet an important part of not hogging up all connection slots.

**Comment by dryajov** [comment]
> I would like to understand the rationale why libp2p connection manager makes a dial fail fast when the maximum allowed peers is reached, while for incoming connections it can await a free slot.
https://github.com/status-im/nim-libp2p/blob/master/libp2p/connmanager.nim#L423 versus https://github.com/status-im/nim-libp2p/blob/master/libp2p/connmanager.nim#L446

> Would it not be better to allow for both to wait? And that way allow the "user" to cancel the dial when it waited long enough.

The main reason is because there is no discernable way of telling weather the connection is waiting for the remote to respond, ie, the remote being slow/malicious vs simply waiting for a connection slot. That's why failing fast _seems_ to be the only way of making that distinction.

> I realize this can be resolved by splitting limits over the inSema and outSema, but still, I think it would be nice to have the waiting behaviour also for outgoing connections.

Overall, this is probably the best way of approaching this issue, this was in fact the reason to have separate counts in the first place.

> Additionally incoming connections allow to be keep added up waiting (not added to the peer count though luckily) because they wait on yet another semaphore here: https://github.com/status-im/nim-libp2p/blob/master/libp2p/switch.nim#L324
Would it not be better to let these fail fast?

The `upgrades` semaphore is to allow controlling in-flight upgrades. The reason to check it first is to prevent connection slots from being taking up before realizing that we need to wait for an upgrade slot to become available, this is important because the counts _can be shared_ across incoming and outgoing connections, as it is the case now. We only care about upgrades for incoming connections, outgoing connections are controlled by the calling code to allow the application to make that decision.

> That logic should probably also go in the connection management as now it is separated and not so visible, yet an important part of not hogging up all connection slots.

Not a bad observation and perhaps you're correct that this is now better suited in the connection manager, tho it's worth considering it in the context of my comment in the previous paragraph.

**Comment by dryajov** [comment]
> The main reason is because there is no discernable way of telling weather the connection is waiting for the remote to respond, ie, the remote being slow/malicious vs simply waiting for a connection slot. That's why failing fast seems to be the only way of making that distinction.

To clarify, in the case of NBC, this would still raise, but after a timeout had elapsed, which has the same effect but with the added timeout delay.

**Comment by kdeme** [comment]
> The main reason is because there is no discernable way of telling weather the connection is waiting for the remote to respond, ie, the remote being slow/malicious vs simply waiting for a connection slot. That's why failing fast _seems_ to be the only way of making that distinction.

Right, that makes sense. I was mostly thinking how it currently doesn't seem to fit that well in the current nbc design. In it its current design it would make sense to await on those dials too and eventually time out on them when no slots, this way previously discovered peers don't get thrown away. But, my thinking here is also flawed as, in this case those nodes would also be added to the `SeenTableTime` in nbc and not redialed until after some time, basically the point you are making of not being able to differentiate. 

> The `upgrades` semaphore is to allow controlling in-flight upgrades. The reason to check it first is to prevent connection slots from being taking up before realizing that we need to wait for an upgrade slot to become available, this is important because the counts _can be shared_ across incoming and outgoing connections, as it is the case now. We only care about upgrades for incoming connections, outgoing connections are controlled by the calling code to allow the application to make that decision.

I'm not sure I understand here what you mean with in-flight upgrades. I do get the point for why the `ConcurrentUpgrades` semaphore is needed. Which was actually my point after that. Initially I had not noticed this semaphore, and I thought that one could take over all connection slots by dialing and then idling for many connections. But that semaphore blocks that, which is why I indicated its importance.

> To clarify, in the case of NBC, this would still raise, but after a timeout had elapsed, which has the same effect but with the added timeout delay.

Yeah, that's clear thanks. It is what I meant with "And that way allow the "user" to cancel the dial when it waited long enough."

**Comment by dryajov** [comment]
> basically the point you are making of not being able to differentiate.

I initially implemented it without the raise and I hit the behaviour with NBC throttling dials, which was puzzling until I realized that it was timing out and marking this peers as seen. It was bad enough that it stopped dialing any peers at all pretty quickly, within minutes.

> I'm not sure I understand here what you mean with in-flight upgrades. 

Right, by in-flight upgrades I mean upgrades that have initiated but not yet completed. There are a number of steps to upgrade a TCP connection (this might be slightly different for other transports, such as quick) and each of this steps might take an arbitrary amount of time. 

It looks something like this:

> dial/accept -> connection -> secure -> identify -> mux -> done

For TCP, until all this steps are completed, we don't have a usable connection, in fact, when calling `dial` it won't return until all of this have completed. The problem comes up with how connection slots are accounted for, we use a connection slot as soon as the transport returns a connection, either thru `dial` or `accept`. This is done so that we prevent an unbound amount of sockets and consequently in-flight connections.

This is also an important attack vector, where an attacker initiates a bunch of dials to a node and then just hangs in one of the steps, for example `identify`. The connection will eventually timeout, but no dials would be possible because all the slots are taken up, this will probably not DDoS the node entirely, but it can definitely degrade its performance pretty badly.

Having a semaphore for accepts will prevent that (mostly), because we don't pick the next connection up until all in-flight accepts are finished.

There is however an issue that we need to fix rather quickly, it is luckily an easy fix - we need to limit our TCP backlog to some sane number, which right now it's set to the chronos default of 100.

What I suspect is going on is that there are a number of connections in the backlog that get immediately picked up by accept, as soon as the previous one finished. IIRC, the backlog is only for incoming connections and as soon as it's full, the server will immediately drop any connection attempts. Having a backlog of 100 means that there are up to 100 ready connections waiting to be picked up by accept and depending on how chronos is handling that (and this is obviously platform specific to add to the mix), it might block dials entirely. This is speculative, but it seems at least plausible and worth trying out, specially given that all it takes is tweaking the parameter when the transport is created. The only problem is that this param isn't exposed on the libp2p tcp transport, so it might require a patch or manual tweaks.

**Comment by kdeme** [comment]
Sorry, forgot to reply here. 
All is pretty clear now, thanks! As I have no concrete suggestions for changes/improvements for now, I think this can be closed.
